### PR TITLE
[BD-46] docs: improve design and usability of Icons page

### PR DESCRIPTION
**Pull request by monteri**
## Description

Features and fixes:
- according to big number of icons it required time to render all the icons. To avoid instant rendering of 2000+ icons `react-window` was used
- search against 2000+ icons is also a heavy operation. To avoid multiple searches `debounce` from `lodash` was used
- preview of the icon is added on the top of the table. If user scrolls to the bottom of the table and clicks on some icon then he is auto-scrolled to the preview.
- added styles
- fix for the SearchField: re-rendering of the `Wrapper` inside the `SearchField` makes the inner components to unmount and mount again that causes problems when writing a text in the input. Added `useCallback` to prevent re-rendering

### Deploy Preview

[Icons page](https://deploy-preview-1378--paragon-openedx.netlify.app)

## Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [x] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [ ] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [x] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [ ] Were your changes tested in the `example` app?
* [x] Is there adequate test coverage for your changes?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please add `wittjeff` and `adamstankiewicz` as reviewers on this PR.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@edx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @monteri! I've created [BLENDED-1314](https://openedx.atlassian.net/browse/BLENDED-1314) to keep track of it in Jira.   More details are on the [BD-46](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3134685276/BD-46+Paragon+Blended+Support+and+Maintenance) project page.


When this pull request is ready, tag your edX technical lead.

**Review by viktorrusakov — COMMENTED** [review]
Left couple nits, otherwise LGTM!

Also in the deploy preview I noticed something weird in the deploy preview
image

But it's not caused by your changes, I see the same issue on the production site too and not just on the Icons page but on all non-component pages...
I'll file a separate ticket for that.

File: www/src/components/IconsTable.jsx [hunk 1]
```diff
@@ -1,27 +1,177 @@
...
-          <h3 className="mb-0 mr-3 flex-grow-1 h5">{iconName}</h3>
-          <code className="bg-light px-2 py-1 rounded">
+          <code
+            role="presentation"
+            className="rounded"
+            onClick={() => copyToClipboard(currentIconImport)}
+          >
             <small>
-              {`import { ${iconName} } from '@edx/paragon/icons';`}
+              {currentIconImport}
```
Comment (adamstankiewicz) [inline www/src/components/IconsTable.jsx:109]: I'm thinking it might be a good idea to disclose that it is possible to copy this import string, without having to stumble on it. Maybe we show an inline/small `ContentCopy` icon to the right of the import statement that is the trigger to copy?

File: www/src/components/IconsTable.jsx [hunk 2]
```diff
@@ -1,27 +1,177 @@
...
-            src={IconComponents[iconName]}
+            key={currentIcon}
+            src={IconComponents[currentIcon]}
           />
-          <h3 className="mb-0 mr-3 flex-grow-1 h5">{iconName}</h3>
-          <code className="bg-light px-2 py-1 rounded">
+          <code
+            role="presentation"
+            className="rounded"
+            onClick={() => copyToClipboard(currentIconImport)}
```
Comment (adamstankiewicz) [inline www/src/components/IconsTable.jsx:106]: Let's add a Segment event to track clicks, e.g. `openedx.paragon.icons-table.selected-icon.copied`, and maybe pass the name of the icon as a `name` property as well?

File: www/src/components/IconsTable.jsx [hunk 3]
```diff
@@ -1,27 +1,177 @@
...
+              {currentIconImport}
             </small>
           </code>
         </div>
-      ))}
-    </div>
+      )}
+      {iconsList.length ? (
+        <div className="pgn-doc__icons-table">
+          <AutoSizer>
```
Comment (adamstankiewicz) [inline www/src/components/IconsTable.jsx:116]: It appears there's a double-scroll situation in which this table can be scrolled as well as the main page viewport. Is there a way we can eliminate the need for the table to be scrollable?

If we're trying to eliminate displaying icons that aren't visible in the viewport, I'd think additional rows of icons would start appearing as the main page scroll gets closer to the end of the page, if that makes sense.

Comment (adamstankiewicz) [inline www/src/components/IconsTable.jsx:116]: When there aren't many icons displayed, it appears there's a lot of excessive/unnecessary whitespace between the end of the icons table and the footer.

image

File: www/src/components/IconsTable.scss [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    margin: 0 auto;
+  }
+}
+
+.pgn-doc-icons-table__grid {
+  margin: 0 auto;
+}
+
+.pgn-doc__icons-table__cell {
+  @include pgn-box-shadow(1, "centered");
```
Comment (adamstankiewicz) [inline www/src/components/IconsTable.scss:22]: With the cells flush together, the box-shadow with adjacent cells kind of conflicts. I wonder if the UI might be a bit cleaner without any shadows/borders, except on hover/focus:

image

File: www/src/components/IconsTable.jsx (hunk 1 above)
Comment (monteri) [reply www/src/components/IconsTable.jsx:109]: Yes, I added `ContentCopy ` to the right and hover styles

File: www/src/components/IconsTable.jsx (hunk 3 above)
Comment (monteri) [reply www/src/components/IconsTable.jsx:116]: I changed to own infinite scroll implementation

Comment (monteri) [reply www/src/components/IconsTable.jsx:116]: Updated to `min-height: 200px`

**Review by adamstankiewicz — APPROVED** [review]
Awesome! I think this came out really cool 😄 Great work!

**Comment by openedx-webhooks** [comment]
@monteri 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/paragon&entry.867055334=https://github.com/openedx/paragon/pull/1378&entry.1484655318=https://github.com/monteri&entry.752974735=2022-06-13+13:10&entry.1917517419=2022-06-24+03:24&entry.2133058324=Yes) so we can improve your experience in the future.
